libgalois/include/katana/PropertyGraph.h

```diff
@@ -202,7 +203,7 @@ class KATANA_EXPORT PropertyGraph {
       return (g->*add_properties_fn)(props);
     }
 
-    Result<void> UpsertProperties(
+    katana::Result<std::vector<katana::fbs::RDGAccessT>> UpsertProperties(
```

Comment:
I would strongly advise against using flatbuffer types in the API. The reason is that if we want to change storage format we would be forced to change the API too. I think there should be a pure C++ wrapper that can be updated to work with a different backing data store if we change storage formats and stop using flatbuffer.

This applies throughout.

Reply:
I would really rather hit the conflict detection machinery. I would prefer passing a mutable transaction object into the function that is used to log the accesses. Returning something here makes it very easy to just throw away the results by accident and not add them to the actual read/write set causing transaction soundness bugs.

Reply:
@yige-hu can you make a `tsuba::Access` class to hold the information for `katana::fbs::RDGAccessT`? If you do that, you don't even need to put the flatbuffer definitions in this repo.

> I would prefer passing a mutable transaction object into the function that is used to log the accesses. Returning something here makes it very easy to just throw away the results by accident

It would be good to make sure the access sets are updated consistently, but I don't see how `void fn(txn_log)` would be any more safe or hard to misuse than `access fn()`.

Reply:
We might prefer `void fn(txn_context)` not because it is safer but because it is easier to extend without changing the callees. For instance if we wanted to add more information about a transaction, we can have the leaf functions update the `txn_context` and all the intervening functions don't have to change their API.

Reply:
The reason I say it's safer is that it's not hard to store access information into the wrong log or not store it at all. And those would create silent unhandled transaction conflicts that could corrupt data. Unless I misunderstand.

Reply:
Aside: with `[[nodiscard]]` you can force users to at least do something with return values, while no one is expected to do anything with their parameters.

I think having a "transaction context" would be good for other reasons (easier to iterate the API). Having an explicit transaction context is a pretty big (initial) change to the API though. The alternative would be some thread-local variable, but that seems worse.

Thoughts?

Reply:
I wasn't implying you would have a parameter you have to do something with. I was saying you would have a transaction context of some kind that the transaction engine would be aware of and would look at when it performed commits or abort points. The normal OO approach to this is to have the transaction object have the mutating operations on it, or pass the transaction object to the mutating operations.

`[[nodiscard]]` helps, for sure. But there are two issues: it doesn't exist in python, and it's still easy to store it in a variable and then drop it, or fail to merge the different access sets together correctly or what not. I think in python we should have an implicit context using `with`. It's conventional in Python. We could implement this based on returned access sets, but it would be much easier to implement using a context object.

Reply:
This is clearly pretty complex, so we can definitely push this off to another PR. Though I don't love planning to have breaking API changes.

Reply:
The standard is to have a transaction context object (`void fn(txn_context)`), which gets rid of the flatbuffer type from the API.
My first instinct would be to insist on that as early as possible, but I'm not sure how that plays with the alpha deadline.
Yige, how long would it take you to add a transaction object?

Reply:
@arthurp @ddn0 I agree with the idea of a transaction context while I was in the middle of coding this PR. I still sent out the PR to collect opinions. I also considered enforcing all related classes to inherit a transactional object base class. But that might be less intuitive compared to passing a transaction context.

@witchel I think it should take around 2-3 days to modify the code changes covered by this PR.

I would support modifying this PR instead of doing it in other PR, so that I don't need to merge all the API changes twice.